lib/ash/changeset/changeset.ex

```diff
@@ -4137,6 +4137,51 @@ defmodule Ash.Changeset do
     end
   end
 
+  @doc """
+  Updates an existing attribute change by applying a function to it.
+
+  This is useful for applying some kind of normalization to the attribute.
+
+  ```elixir
+  Ash.Changeset.update_change(changeset, :serial, &String.downcase/1)
+  ```
+
+  The update function gets called with the value already cast to the correct type.
+
+  ```elixir
+  changeset
+  |> Ash.Changeset.change_attribute(:integer_attribute, "3")
+  |> Ash.Changeset.update_change(:integer_attribute, fn x -> x + 1 end)
+  ```
+
+  ## `nil` handling
+
+  If the attribute has `allow_nil?: false`, the update function is guaranteed not to be called if the
+  attribute value is `nil`, even if it is set to `nil` explicitly.
+
+  If the attribute has `allow_nil?: true` the update function must consider `nil` as a possible input
+  if it is passed explicitly.
+  """
+  @spec update_change(t(), atom, (any -> any)) :: t()
+  def update_change(changeset, attribute, fun) do
+    case Ash.Changeset.fetch_change(changeset, attribute) do
```

Comment:
I think we should not call the function if the changeset is invalid as well. That way if a validation is applied on a value that doesn't pass, you don't try to update it. For instance:

```elixir
accept [:...]
validate compare(points, greater_than: 0)
change update_change(:points, &(&1 - 1))
```

Reply:
This doesn't work (e.g. for `allow_nil` validations) if it gets called like this:

```elixir
    test "does not call the update function on explicit `nil` attribute with `allow_nil?: false`" do
      Tag
      |> Ash.Changeset.new()
      |> Ash.Changeset.change_attribute(:name, nil)
      |> Ash.Changeset.update_change(:name, fn _ -> raise "I should not be called" end)
      |> Ash.Changeset.for_create(:create)
    end
```

this actually raises (and `:name` has `allow_nil?: false`)

It works moving the call to `update_change` downwards, but then the implementation has to use `force_change_attribute`. I'm not sure if the function should be designed to be used before or after `Changeset.for_*`, or both.

Reply:
Yeah, I think if we do everything you’re doing *and* check for validity that would do what we want. It will work best-effort before the action, and fully during the action.

Reply:
I've tried to do that, but it's still brittle (for example, action validations run before `update_change` if they're before it _but_ global validations run after, so the update function can be called with a value that fails a global validation).

Given this not so obvious pieces of behavior, I think it's best to drop the function in the `Changeset` module and just provide a builtin that gets called in a `before_action` hook, like it's already done in, e.g., [Increment](https://github.com/ash-project/ash/blob/main/lib/ash/resource/change/increment.ex)

Reply:
global validations should happen after action-specific changes.

Reply:
I think we can do it in a before action hook 👍 only issue is that it could throw users for a loop expecting it to be updated.